### ISSUE TITLE
Regression in NoReturn

### ISSUE THREAD

**Issue by 0xjc** [microsoft/pyright#3531]
Minimal example:

```
# pyright: strict

from typing import NoReturn, Optional


def raise_error() -> NoReturn:
    raise ValueError("hi")


def g(a: int) -> int:
    x: Optional[int] = None
    if a > 0:
        x = a
    if x is None:
        raise_error()
    return x
```

Error:
```
  /tmp/pytest.py:16:12 - error: Expression of type "int | None" cannot be assigned to return type "int"
    Type "int | None" cannot be assigned to type "int"
      Type "None" cannot be assigned to type "int" (reportGeneralTypeIssues)
```

- pyright 1.1.251
- The lines `if a > 0: x = a` are not required to repro, they just make the example more realistic
- If `raise ValueError("hi")` is used directly instead of `raise_error()`, the type check passes
- If `assert x is not None` is added before `return x`, the type check passes

**Comment by ottumm** [comment]
Another example that I think illustrates the same bug:

```python
# pyright: strict

import sys


def foo(a: int | None):
    if a is None:
        sys.exit(1)

    return a + 2
```

```
/src/test.py
  /src/test.py:10:12 - error: Operator "+" not supported for "None" (reportOptionalOperand)
  /src/test.py:10:12 - error: Return type is unknown (reportUnknownVariableType)
  /src/test.py:6:5 - error: Return type is unknown (reportUnknownParameterType)
```

**Comment by rslinckx** [comment]
Same here with a flask common pattern:
```
from flask import abort, Flask
app = Flask(__name__)
@app.route('/')
def index():
    obj: Obj|None = Obj.query.first()
    if not obj:
        abort(404)
    return obj.foo # <- here obj is still Obj|None where it should be Obj
```
This worked with 1.1.249

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release. I've added new test cases based on the code samples above to prevent similar regressions from occurring in the future.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.252, which I just published. It will also be included in the next release of pylance.

**Comment by rslinckx** [comment]
@erictraut  There is still something weird:

```
from flask import Flask, abort

app = Flask(__name__)


def get_something_or_none() -> str | None:
    return None


@app.route("/")
def index():
    x = get_something_or_none()
    if not x:
        abort(404)

    def foo() -> str:
        return x.upper() # <- upper is not a known member of "None"

    return x.upper() # <- upper is not a known member of "None"
```

while this version type-checks properly (I just removed the inline foo() definition):

```
from flask import Flask, abort

app = Flask(__name__)


def get_something_or_none() -> str | None:
    return None


@app.route("/")
def index():
    x = get_something_or_none()
    if not x:
        abort(404)

    return x.upper()
```

**Comment by erictraut** [comment]
Could you please open another issue? This one is different, and it's best to track it separately.

**Comment by rslinckx** [comment]
Yes, sure: https://github.com/microsoft/pyright/issues/3551
